ambry-tools/src/main/java/com.github.ambry/clustermap/HelixBootstrapUpgradeTool.java

```diff
@@ -0,0 +1,470 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.clustermap;
+
+import com.github.ambry.config.ClusterMapConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.tools.util.ToolUtils;
+import com.github.ambry.utils.Utils;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Properties;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import joptsimple.ArgumentAcceptingOptionSpec;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.manager.zk.ZKHelixAdmin;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.LeaderStandbySMD;
+import org.apache.helix.model.builder.AutoModeISBuilder;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+
+/**
+ * This tool takes the hardware layout, partition layout and the Zk hosts information json files as input,
+ * and updates the ZK hosts with the contents of the layout files. It adds all partitions and hosts that have not
+ * previously been added (so, initially this will bootstrap the cluster information and on an ongoing basis, this can
+ * add new nodes and partitions).
+ *
+ * The existing hardware and partition layout json files will be read in as is.
+ *
+ * The ZkLayoutPath argument containing the Zk hosts information in each datacenter should be a json of the
+ * following example form:
+ *
+ * {
+ *   "zkHosts" : [
+ *     {
+ *       "datacenter":"dc1",
+ *       "hostname":"abc.example.com",
+ *       "port":"2199",
+ *     },
+ *     {
+ *       "datacenter":"dc2",
+ *       "hostname":"def.example.com",
+ *       "port":"2300",
+ *     }
+ *   ]
+ * }
+ *
+ * This tool should be run from an admin node that has access to the nodes in the hardware layout. The access is
```

Comment:
do you think we should word it as "this tool should be run (in every datacenter) from an admin node..."